src/hotspot/share/opto/loopnode.cpp

```diff
@@ -1109,6 +1116,184 @@ bool PhaseIdealLoop::create_loop_nest(IdealLoopTree* loop, Node_List &old_new) {
+  }
+
+  if (!known_short_running_loop && !profile_short_running_loop) {
+    return false;
+  }
+
+  Node* limit = head->limit();
+  Node* init = head->init_trip();
+
+  Node* new_limit;
+  if (stride_con > 0) {
+    new_limit = SubNode::make(limit, init, bt);
+  } else {
+    new_limit = SubNode::make(init, limit, bt);
+  }
+  register_new_node(new_limit, entry_control);
+
+  PhiNode* phi = head->phi()->as_Phi();
+  const Type* new_phi_t = TypeInt::INT;
+  if (profile_short_running_loop) {
+    // Add a short_limit predicate. It's the last predicate when coming from the loop because a cast that's control
+    // dependent on the short_limit predicate is added to narrow the limit and future predicates may be dependent on the
+    // new limit (so have to be between the loop and short_limit predicate). The current limit could, itself, be
+    // dependent on an existing predicate. Clone parse predicates below existing predicates to get proper ordering of
+    // predicates when coming from the loop: future predicates, short_limit predicate, existing predicates.
+    const Predicates predicates(entry_control);
+    const PredicateBlock* short_running_loop_predicate_block = predicates.short_running_loop_predicate_block();
+    if (!short_running_loop_predicate_block->has_parse_predicate()) { // already trapped
+      return false;
+    }
+    const PredicateBlock* predicate_block = predicates.loop_predicate_block();
+    ParsePredicateSuccessProj* parse_predicate_proj = short_running_loop_predicate_block-> parse_predicate_success_proj();
+    Node* ctrl = entry_control;
+    ctrl = clone_parse_predicate(parse_predicate_proj, ctrl,
+                                                    Deoptimization::Reason_short_running_loop, true);
+    Node* short_running_loop_ctrl = ctrl;
+    if (predicate_block->has_parse_predicate()) {
+      parse_predicate_proj = predicate_block->parse_predicate_success_proj();
+      ctrl = clone_parse_predicate(parse_predicate_proj, ctrl, Deoptimization::Reason_predicate,
+                                                      true);
+      Unique_Node_List list;
+      get_template_assertion_predicates(parse_predicate_proj, list);
+      clone_assertion_predicates(loop, list, ctrl->in(0)->as_ParsePredicate());
+    }
+    const PredicateBlock* profiled_predicate_block = predicates.profiled_loop_predicate_block();
+    if (profiled_predicate_block->has_parse_predicate()) {
+      parse_predicate_proj = profiled_predicate_block->parse_predicate_success_proj();
+      ctrl = clone_parse_predicate(parse_predicate_proj, ctrl,
+                                                      Deoptimization::Reason_profile_predicate, true);
+      Unique_Node_List list;
+      get_template_assertion_predicates(parse_predicate_proj, list);
+      clone_assertion_predicates(loop, list, ctrl->in(0)->as_ParsePredicate());
+    }
```

Comment:
You may want to talk with @chhagedorn to see if this cannot be done with less code-duplication.
Also: where are the `Unique_Node_List` allocated from / deallocated?